dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/Application/MessageExtensions.cs

```diff
@@ -0,0 +1,735 @@
+        {
+            Verify.ParamNotNull(commandId);
+            Verify.ParamNotNull(handler);
+            RouteSelector routeSelector = CreateTaskSelector((string input) => string.Equals(commandId, input), SUBMIT_ACTION_INVOKE_NAME, "edit");
+            return OnBotMessagePreviewEdit(routeSelector, handler);
+        }
+
+        /// <summary>
+        /// Registers a handler to process the 'edit' action of a message that's being previewed by the
+        /// user prior to sending.
+        /// </summary>
+        /// <param name="commandIdPattern">ID of the command to register the handler for.</param>
+        /// <param name="handler">Function to call when the command is received.</param>
+        /// <returns>The application for chaining purposes.</returns>
+        public Application<TState, TTurnStateManager> OnBotMessagePreviewEdit(Regex commandIdPattern, BotMessagePreviewEditHandler<TState> handler)
+        {
+            Verify.ParamNotNull(commandIdPattern);
+            Verify.ParamNotNull(handler);
+            RouteSelector routeSelector = CreateTaskSelector((string input) => commandIdPattern.IsMatch(input), SUBMIT_ACTION_INVOKE_NAME, "edit");
+            return OnBotMessagePreviewEdit(routeSelector, handler);
+        }
+
+        /// <summary>
+        /// Registers a handler to process the 'edit' action of a message that's being previewed by the
+        /// user prior to sending.
+        /// </summary>
+        /// <param name="routeSelector">ID of the command to register the handler for.</param>
+        /// <param name="handler">Function to call when the command is received.</param>
+        /// <returns>The application for chaining purposes.</returns>
+        public Application<TState, TTurnStateManager> OnBotMessagePreviewEdit(RouteSelector routeSelector, BotMessagePreviewEditHandler<TState> handler)
+        {
+            RouteHandler<TState> routeHandler = async (ITurnContext turnContext, TState turnState, CancellationToken cancellationToken) =>
+            {
+                MessagingExtensionAction? messagingExtensionAction;
+                if (!string.Equals(turnContext.Activity.Type, ActivityTypes.Invoke, StringComparison.OrdinalIgnoreCase)
+                    || !string.Equals(turnContext.Activity.Name, SUBMIT_ACTION_INVOKE_NAME)
+                    || (messagingExtensionAction = Application<TState, TTurnStateManager>.GetInvokeValue<MessagingExtensionAction>(turnContext.Activity)) == null
+                    || !string.Equals(messagingExtensionAction.BotMessagePreviewAction, "edit"))
+                {
+                    throw new TeamsAIException($"Unexpected MessageExtensions.OnBotMessagePreviewEdit() triggered for activity type: {turnContext.Activity.Type}");
+                }
+
+                MessagingExtensionActionResponse result = await handler(turnContext, turnState, messagingExtensionAction.BotActivityPreview.FirstOrDefault(), cancellationToken);
```

Comment:
`FirstOrDefault` may return null, is null error handling required?

Reply:
Not sure which is better: passing `null`, or passing an empty activity, i.e. `new Activity()`. Any idea?

Reply:
If it really could be null, declare handler's parameter as nullable let user to know it could be null in xml doc.

Reply:
Verified that `messagingExtensionAction.BotActivityPreview` will never be empty. So we can keep it simple as `messagingExtensionAction.BotActivityPreview[0]`.